### PR TITLE
Improve performance of String.concat for seq of type array or list

### PR DESCRIPTION
**Pull request by abelbraaksma**
This follows the findings and timings shown in this comment: https://github.com/dotnet/fsharp/issues/9390#issuecomment-641553171

Basically: special-casing the input `seq` for array gives 2x perf gain and less 2-3x less mem use, and special-casing for `list` gives 20% gain and 30% less mem and GC pressure.

@KevinRansom: this is the one we discussed that shows no improvement by using array-based manipulation if the input is a `seq`, and since `String.Join` uses buffered-SB internally, that may well be the reason it already performed so well.

It's also a showcase of the internal `ArrayEnumerator` of arrays-turned-into-sequences, however fast it is, still adding a lot of time for the indirection, hence the success of the special-casing of array. Try this new method with `Array.toSeq x` and with implicit `x :> seq<_>` and watch the difference (a factor 2).

This PR is dependent on https://github.com/dotnet/fsharp/pull/9481.

### CODE REVIEW

File: src/fsharp/FSharp.Core/string.fs [hunk 1]
```diff
@@ -8,13 +8,34 @@ namespace Microsoft.FSharp.Core
...
     module String =
+        [<CompiledName("Length")>]
+        let length (str:string) = if isNull str then 0 else str.Length
+
+        let private concatArray sep (strings: string []) =
+            match length sep with
+            | 0 -> String.Concat strings
+            // following line should be used when this overload becomes part of .NET Standard (it's only in .NET Core)
+            //| 1 -> String.Join(sep.[0], strings, 0, strings.Length)
+            | _ -> String.Join(sep, strings, 0, strings.Length)
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/string.fs:24]: Typically these are done as inner functions instead of private ones. Could you make that change?

Comment (abelbraaksma) [reply src/fsharp/FSharp.Core/string.fs]: Sure, done, fixed: https://github.com/dotnet/fsharp/pull/9483/commits/d7c95e00055b35484c7d2ae6f789fd3a5fb0f138

**Comment by abelbraaksma** [comment]
Since we now have three execution paths (for list, seq and array), I changed the tests to execute all three paths for the existing tests.

**Comment by abelbraaksma** [comment]
Strange, it keeps saying "Changes requested" even though all requests have been resolved. Maybe I'm doing something wrong...

**Comment by KevinRansom** [comment]
@abelbraaksma , the requester has to acknowledge that they have been handled I believe.

File: tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/StringModule.fs [hunk 2]
```diff
@@ -23,30 +23,26 @@ type StringModule() =
...
+            let r1 = f (List.toSeq arg)
+            Assert.AreEqual(expected, r1)
+
+            let r2 = f (List.toArray arg)
+            Assert.AreEqual(expected, r2)
+
+            let r3 = f arg
+            Assert.AreEqual(expected, r3)
+
+        do execTest (String.concat null) "foo" ["foo"]
```
Comment (KevinRansom) [inline tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/StringModule.fs:37]: Can we use words other than foo and bar.  They cause eyes to be raised.  I expect we have them in the codebase, but I imagine we will get rid of them eventually.

Comment (abelbraaksma) [reply tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/StringModule.fs:37]: No problem. These are the tests as they were, I didn't change them, but I personally prefer something that's easier to check visually. I'll update.

Comment (KevinRansom) [reply tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/StringModule.fs:37]: @abelbraaksma , thanks for taking care of that mate ... eradicating age old nonsense is a lifetimes work.

Comment (abelbraaksma) [reply tests/FSharp.Core.UnitTests/FSharp.Core/Microsoft.FSharp.Collections/StringModule.fs]: @KevinRansom, done in https://github.com/dotnet/fsharp/pull/9483/commits/16103cc38f251696cf96f6cda2477e1833277de1. No "foo", "bar" etc anymore.

**Comment by abelbraaksma** [comment]
All requested changes have been implemented and build is green. @cartermp / @KevinRansom, is it ok as-is?

**Comment by abelbraaksma** [comment]
@cartermp, you merged master into this, I think this is now ready?

**Comment by cartermp** [comment]
@abelbraaksma I think this is ready, but this will require one other review (preferably @KevinRansom since he previously requested changes)

**Comment by abelbraaksma** [comment]
Thanks @kevinransom!

**Pull request merged**